---
title: UCell signature enrichment analysis
author: "M. Andreatta and S. Carmona"
date: "06/04/2021"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'UCell_matrix_vignette.html'))})
---

```{r setup, echo=FALSE, message=F, warning=F, results=F}
#install.packages("rmdformats")
#Template markdown setup
library(knitr)
library(rmdformats)
library(formatR)
library(renv)
#renv::restore()

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=TRUE,
               cache.lazy=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev='png')
opts_knit$set(width=75)

```

In this demo, we will apply UCell to evaluate gene signatures in a set of CD45+ single cells from 48 tumor biopsies of melanoma patients, sequenced using the Smart-seq2 protocol [Sade-Feldman et al. (2018) Cell](https://pubmed.ncbi.nlm.nih.gov/30388456/).

## Installation

```{r message=F, warning=F, results=F}
remotes::install_github("carmonalab/UCell")

library(UCell)
library(Matrix)
set.seed(123)
```

## scRNA-seq data preparation

Download the gene expression matrix for this study from Gene Expression Omnibus - [GSE120575](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120575)

```{r  message=F, warning=F, results=F}
BiocManager::install("GEOquery")
library(GEOquery)

cached.object <- "SadeFeldman.expmat.rds"
if (!file.exists(cached.object)) {
  geo_acc <- "GSE120575"
  gse <- getGEO(geo_acc)
  
  series <- paste0(geo_acc, "_series_matrix.txt.gz")
  getGEOSuppFiles(geo_acc)
  
  ## Load expression matrix and metadata
  exp.mat <- read.delim(sprintf("%s/GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt.gz", 
                                geo_acc), header = F, sep = "\t")
  genes <- exp.mat[c(-1, -2), 1]
  cells <- as.vector(t(exp.mat[1, 2:16292]))
  samples <- as.factor(t(exp.mat[2, 2:16292]))
  
  exp.mat <- exp.mat[c(-1, -2), 2:16292]
  colnames(exp.mat) <- cells
  rownames(exp.mat) <- genes
  
  #Make sparse
  exp.mat <- Matrix::Matrix(as.matrix(exp.mat), sparse = T)
  saveRDS(exp.mat, "SadeFeldman.expmat.rds")
}
```

Alternatively, get the same matrix already processed in sparse format (much faster to load) at [THIS LINK](https://drive.switch.ch/index.php/s/JCAWvnlxCmOvG42)

And then load it using:
```{r}
exp.mat <- readRDS("SadeFeldman.expmat.rds")
```

## Define gene signatures

Here we define some simple gene sets for immune cells based on the "Human Cell Landscape" signatures [Han et al. (2020) Nature](https://www.nature.com/articles/s41586-020-2157-4)
```{r}
set.seed(123)
signatures <- list(Macrophage = c("CTSB","C1QB","LAPTM5","TYROBP","PSAP","C1QA","HLA-DRA","CTSD","NPC2","FCER1G"),
    Tcell = c("CD3D","CD3E","CD3G","CD4","CD2","CD7","TRAC","LAT"),
    Bcell = c("MS4A1","CD79A","CD79B","CD19","BANK1","BCL11A","CD22","NCF1C"),
    Neutrophil = c("SRGN","AZU1","MPO","MS4A3","RNASE2","ELANE","DEFA4","CTSG","PRTN3","DEFA3"),
    Myeloid_cell = c("CD14","CYP17A1","CSF1R","FCER1G","SPI1"),
    NKcell = c("FCGR3A","KLRF1","NCR1","FGFBP2","FCGR3A")
              )
```

## Run UCell

Run `ScoreSignatures_UCell` and get directly signature scores for all cells
```{r}
u.scores <- ScoreSignatures_UCell(exp.mat,features=signatures)
u.scores[1:8,1:3]
```

Show the distribution of predicted scores
```{r fig.height=3, fig.width=7}
library(reshape2)
library(ggplot2)
melted <- melt(u.scores)
colnames(melted) <- c("Cell","Signature","Uscore")
p <- ggplot(melted, aes(x=Signature, y=Uscore)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1) + theme_bw() + theme(axis.text.x=element_blank())
p
```

## Pre-calculating gene rankings

The time- and memory-demanding step in UCell is the calculation of gene rankings for each individual cell. If we plan to experiment with signatures, editing them or adding new cell subtypes, it is possible to pre-calculate the gene rankings once and for all and then apply new signatures over these pre-calculated ranks. Run the `StoreRankings_UCell` function to pre-calculate gene rankings over a dataset:
```{r}
set.seed(123)
ranks <- StoreRankings_UCell(exp.mat)
ranks[1:5,1:5]
```

Then, we can apply our signature set, or any other new signature to the pre-calculated ranks. The calculations will be considerably faster.
```{r fig.height=3, fig.width=7}
set.seed(123)
u.scores.2 <- ScoreSignatures_UCell(features=signatures, precalc.ranks = ranks)

melted <- melt(u.scores.2)
colnames(melted) <- c("Cell","Signature","Uscore")
p <- ggplot(melted, aes(x=Signature, y=Uscore)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1) + theme_bw() + theme(axis.text.x=element_blank())
p
```

```{r fig.height=3, fig.width=5}

new.signatures <- list(Mast.cell = c("TPSAB1","TPSB2","CPA3","SRGN","RGS2","RGS1","NFKBIA","GLUL","VIM","ANXA1"),
                       Erythroid.cell = c("HBA2","HBG2","HBA1","HBB","HBG1","AHSP","ALAS2","SLC25A37","HBM"))

u.scores.3 <- ScoreSignatures_UCell(features=new.signatures, precalc.ranks = ranks)
melted <- melt(u.scores.3)
colnames(melted) <- c("Cell","Signature","Uscore")
p <- ggplot(melted, aes(x=Signature, y=Uscore)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1) + theme_bw() + theme(axis.text.x=element_blank())
p

```

## Visualizing results

We can see how the signatures distribute on low-dimensional representations of the data, for instance by creating a Seurat object from the expression matrix.
```{r}
library(Seurat)
seurat.object <- CreateSeuratObject(counts = exp.mat, project = "SadeFeldman")

seurat.object <- AddMetaData(seurat.object, metadata = as.data.frame(u.scores))
head(seurat.object@meta.data)
```

Generate PCA and UMAP embeddings
```{r}
seurat.object <- NormalizeData(seurat.object, verbose = FALSE)
seurat.object <- FindVariableFeatures(seurat.object, selection.method = "vst", nfeatures = 1000, verbose = FALSE)
  
seurat.object <- ScaleData(seurat.object)
seurat.object <- RunPCA(seurat.object, features = seurat.object@assays$RNA@var.features, verbose = FALSE) 
seurat.object <- RunUMAP(seurat.object, reduction = "pca", dims = 1:20, seed.use=123, verbose = FALSE)
```


```{r fig.height=7, fig.width=12}
signature.names <- paste0(names(signatures), "_UCell")
FeaturePlot(seurat.object, reduction = "umap", features = signature.names, ncol=3, order=T)
```

UCell also implements a wrapper to interact directly with Seurat (see function `AddModuleScore_UCell`). See [THIS VIGNETTE](https://carmonalab.github.io/UCell/UCell_Seurat_vignette.html) for an application of UCell to PBMCs stored in a Seurat object.

## Multi-core processing

If your machine has multi-core capabilities and enough RAM, running UCell in parallel can speed up considerably your analysis. In this example we will use 4 parallel cores:

```{r fig.height=3, fig.width=7}
u.scores <- ScoreSignatures_UCell(exp.mat,features=signatures, ncores=4)

melted <- melt(u.scores)
colnames(melted) <- c("Cell","Signature","Uscore")
p <- ggplot(melted, aes(x=Signature, y=Uscore)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1) + theme_bw() + theme(axis.text.x=element_blank())
p
```

## Further reading

To see how UCell can interact directly with Seurat see [THIS VIGNETTE](https://carmonalab.github.io/UCell/UCell_Seurat_vignette.html)

The code and the package are available at the UCell [GitHub repository](https://github.com/carmonalab/UCell)



