---
title: UCell signature enrichment analysis
author: "M. Andreatta and S. Carmona"
date: "07/02/2022"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'UCell_matrix_vignette.html'))})
---

```{r setup, echo=FALSE, message=F, warning=F, results=F}
#install.packages("rmdformats")
#Template markdown setup
library(knitr)
library(rmdformats)
library(formatR)
library(renv)
#renv::restore()

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=TRUE,
               cache.lazy=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev='png')
opts_knit$set(width=75)

```

In this demo, we will apply UCell to evaluate gene signatures in single-cell dataset from 31 tumor biopsies of melanoma patients, sequenced using the Smart-seq2 protocol [Jerby-Arnon et al. (2018) Cell](https://pubmed.ncbi.nlm.nih.gov/30388455/).

## Installation

```{r message=F, warning=F, results=F}
remotes::install_github("carmonalab/UCell")
library(UCell)
set.seed(123)
```

## scRNA-seq data preparation

Download the gene expression matrix for this study from Gene Expression Omnibus - [GSE115978](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115978)

```{r message=F, warning=F, results=F}
library(GEOquery)
cached.object <- "demodata.rds"

if (!file.exists(cached.object)) {
  geo_acc <- "GSE115978"
  options(timeout = max(1000, getOption("timeout")))
  
  gse <- getGEO(geo_acc)
  getGEOSuppFiles(geo_acc)
  
  exp.mat <- read.csv(sprintf("%s/GSE115978_counts.csv.gz", geo_acc), header=T, row.names=1, sep=",")
  
  saveRDS(exp.mat, cached.object)
} else {
  exp.mat <- readRDS(cached.object)
}
```

## Define gene signatures

Here we define some simple gene sets based on the "Human Cell Landscape" signatures [Han et al. (2020) Nature](https://www.nature.com/articles/s41586-020-2157-4). You may edit existing signatures, or add new one as elements in a list.
```{r}
signatures <- list(
    Immune = c("PTPRC"),
    Macrophage = c("CTSB","C1QB","LAPTM5","TYROBP","PSAP","C1QA","HLA-DRA","CTSD","NPC2","FCER1G"),
    Tcell = c("CD3D","CD3E","CD3G","CD2"),
    Bcell = c("MS4A1","CD79A","CD79B","CD19","BANK1"),
    Myeloid_cell = c("CD14","LYZ","CSF1R","FCER1G","SPI1","LCK-"),
    Stromal = c("MMP2","COL1A1","COL1A2","COL3A1","LUM","DCN")
    )
```

## Run UCell

Run `ScoreSignatures_UCell` and get directly signature scores for all cells
```{r}
u.scores <- ScoreSignatures_UCell(exp.mat,features=signatures)
u.scores[1:8,1:2]
```

Show the distribution of predicted scores
```{r fig.height=3, fig.width=7}
library(reshape2)
library(ggplot2)
melted <- reshape2::melt(u.scores)
colnames(melted) <- c("Cell","Signature","UCell_score")
p <- ggplot(melted, aes(x=Signature, y=UCell_score)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1, outlier.size=0) + theme_bw() + theme(axis.text.x=element_blank())
p
```

## Pre-calculating gene rankings

The time- and memory-demanding step in UCell is the calculation of gene rankings for each individual cell. If we plan to experiment with signatures, editing them or adding new cell subtypes, it is possible to pre-calculate the gene rankings once and for all and then apply new signatures over these pre-calculated ranks. Run the `StoreRankings_UCell` function to pre-calculate gene rankings over a dataset:
```{r}
set.seed(123)
ranks <- StoreRankings_UCell(exp.mat)
ranks[1:5,1:5]
```

Then, we can apply our signature set, or any other new signature to the pre-calculated ranks. The calculations will be considerably faster.
```{r fig.height=3, fig.width=7}
set.seed(123)
u.scores.2 <- ScoreSignatures_UCell(features=signatures, precalc.ranks = ranks)

melted <- reshape2::melt(u.scores.2)
colnames(melted) <- c("Cell","Signature","UCell_score")
p <- ggplot(melted, aes(x=Signature, y=UCell_score)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1, outlier.size = 0) + theme_bw() + theme(axis.text.x=element_blank())
p
```

```{r fig.height=3, fig.width=5}

new.signatures <- list(Mast.cell = c("TPSAB1","TPSB2","CPA3","SRGN","RGS2","RGS1","NFKBIA","GLUL","VIM","ANXA1"),
                       Erythroid.cell = c("HBA2","HBG2","HBA1","HBB","HBG1","AHSP","ALAS2","SLC25A37","HBM"))

u.scores.3 <- ScoreSignatures_UCell(features=new.signatures, precalc.ranks = ranks)
melted <- reshape2::melt(u.scores.3)
colnames(melted) <- c("Cell","Signature","UCell_score")
p <- ggplot(melted, aes(x=Signature, y=UCell_score)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1, outlier.size=0) + theme_bw() + theme(axis.text.x=element_blank())
p

```

## Multi-core processing

If your machine has multi-core capabilities and enough RAM, running UCell in parallel can speed up considerably your analysis. In this example we will use 4 parallel cores:

```{r fig.height=3, fig.width=7}
u.scores <- ScoreSignatures_UCell(exp.mat,features=signatures, ncores=4)

melted <- reshape2::melt(u.scores)
colnames(melted) <- c("Cell","Signature","Uscore")
p <- ggplot(melted, aes(x=Signature, y=Uscore)) + geom_violin(aes(fill=Signature), scale = "width") +
       geom_boxplot(width=0.1, outlier.size=0) + theme_bw() + theme(axis.text.x=element_blank())
p
```


## Interacting with SingleCellExperiment or Seurat

### UCell + SingleCellExperiment

The function `ScoreSignatures_UCell()` allows operating directly on `sce` objects. UCell scores are returned in a altExp object (`altExp(sce, 'UCell'`))
```{r message=F, warning=F}
library(SingleCellExperiment)

sce <- SingleCellExperiment(list(counts=exp.mat))
sce <- ScoreSignatures_UCell(sce, features=signatures, assay = 'counts', name=NULL, ncores=4)
altExp(sce, 'UCell')
```

Dimensionality reduction and visualization
```{r}
library(scater)
library(patchwork)
#PCA
sce <- logNormCounts(sce)
sce <- runPCA(sce, scale=T, ncomponents=20)

#UMAP
set.seed(1234)
sce <- runUMAP(sce, dimred="PCA")
```

Visualize UCell scores on low-dimensional representation (UMAP)
```{r fig.height=9, fig.width=15}
pll <- lapply(names(signatures), function(x) {
    plotUMAP(sce, colour_by = x, by_exprs_values = "UCell", text_size=10)
})
wrap_plots(pll)
```

### UCell + Seurat

The function `AddModuleScore_UCell()` allows operating directly on Seurat objects. UCell scores are returned as metadata columns in the Seurat object. To see how this function differs from Seurat's own `AddModuleScore()` (not based on per-cell ranks) see [this vignette](https://carmonalab.github.io/UCell_demo/UCell_Seurat_vignette.html)
```{r}
library(Seurat)
seurat.object <- CreateSeuratObject(counts = exp.mat, project = "JerbyArnon")
seurat.object <- AddModuleScore_UCell(seurat.object, features=signatures, name=NULL, ncores=4)
head(seurat.object@meta.data)
```

Generate PCA and UMAP embeddings
```{r message=F, warning=F, results=F}
seurat.object <- NormalizeData(seurat.object)
seurat.object <- FindVariableFeatures(seurat.object, selection.method = "vst", nfeatures = 500)
  
seurat.object <- ScaleData(seurat.object)
seurat.object <- RunPCA(seurat.object, features = seurat.object@assays$RNA@var.features, npcs = 20) 
seurat.object <- RunUMAP(seurat.object, reduction = "pca", dims = 1:20, seed.use=123)
```

Visualize UCell scores on low-dimensional representation (UMAP)
```{r fig.height=7, fig.width=12}
FeaturePlot(seurat.object, reduction = "umap", features = names(signatures), ncol=3, order=T)
```


## Further reading

To see how UCell can interact directly with Seurat to analyse human PBMCs see [THIS VIGNETTE](https://carmonalab.github.io/UCell_demo/UCell_Seurat_vignette.html)

The code and the package are available at the UCell [GitHub repository](https://github.com/carmonalab/UCell); more demos available at [UCell demo repository](https://github.com/carmonalab/UCell_demo)



